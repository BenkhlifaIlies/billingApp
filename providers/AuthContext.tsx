import React from "react";
import { useState, createContext, useContext } from "react";
const useAuth = (initial: string|null) => useState<string|null>(initial);

const AuthContext = createContext<ReturnType<typeof useAuth> | null>(
  null,
);

export const useAuthContext = () => useContext(AuthContext)!;

export function AuthProvider({ children }: { children: React.ReactNode }) {
  return (
    <AuthContext.Provider value={useAuth(null)}>
      {children}
    </AuthContext.Provider>
  );
}
