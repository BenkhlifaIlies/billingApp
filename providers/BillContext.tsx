import React from "react";
import { useState, createContext, useContext } from "react";
const useBill = (initial: number) => useState<number>(initial);

const BillContext = createContext<ReturnType<typeof useBill> | null>(
  null,
);

export const useBillContext = () => useContext(BillContext)!;

export function BillProvider({ children }: { children: React.ReactNode }) {
  return (
    <BillContext.Provider value={useBill(0.05)}>
      {children}
    </BillContext.Provider>
  );
}
