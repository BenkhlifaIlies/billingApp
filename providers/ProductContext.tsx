import React from "react";
import { useState, createContext, useContext } from "react";
import { Product } from "../constants/Product";

const useProducts = (initial: Product[]) => useState<Product[]>(initial);

const ProductContext = createContext<ReturnType<typeof useProducts> | null>(
  null,
);

export const useProductContext = () => useContext(ProductContext)!;

export function ProductProvider({ children }: { children: React.ReactNode }) {
  return (
    <ProductContext.Provider value={useProducts([])}>
      {children}
    </ProductContext.Provider>
  );
}
