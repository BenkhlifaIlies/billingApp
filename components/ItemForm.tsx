import React from "react";
import {
  ActivityIndicator,
  StyleSheet,
  TextInput,
  View,
  Text,
} from "react-native";
import { Formik } from "formik";
import { Picker } from "@react-native-picker/picker";
import * as yup from "yup";
import { appleBlue, lightGrey } from "../constants/DesignColors";
import { useProductContext } from "../providers/ProductContext";
import { useBillContext } from "../providers/BillContext";
import { MAXQUANTITY } from "../constants/Calculattions";

const validationSchema = yup.object().shape({
  productLabel: yup
    .string()
    .required()
    .label("Product")
    .min(4, "Product's name is required to be at least 4 characters long"),
  productPrice: yup
    .number()
    .required()
    .label("Price")
    .test(
      "Is positive?",
      "The Price must be greater than 0",
      (value: any) => value > 0,
    )
    .typeError("The Price must be a positive number"),
  quantity: yup
    .string()
    .required()
    .label("Quantity")
    .test(
      "Is positive?",
      "Quantity must be greater than or equal to  1 and inferieur to" +
        MAXQUANTITY,
      (value: any) => value >= 1 && value < MAXQUANTITY,
    ),
});

const ItemForm = () => {
  const [productList, setProductList] = useProductContext();
  const [tax, setTax] = useBillContext();

  const handleInsertion = (
    productLabel: string,
    productPrice: number,
    quantity: number,
  ) => {
    const newArray = productList;
    try {
      newArray.unshift({
        productLabel: productLabel,
        productPrice: productPrice,
        quantity: quantity,
      });
      setProductList([...newArray]);
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <View style={styles.Screenontainer}>
       <Formik
        initialValues={{ productLabel: "", productPrice: "", quantity: "" }}
        onSubmit={(values, actions) => {
          handleInsertion(
            values.productLabel,
            Number(values.productPrice),
            Number(values.quantity),
          );
          actions.resetForm();
          setTimeout(() => {
            actions.setSubmitting(false);
          }, 200);
        }}
        validationSchema={validationSchema}>
        {(formikProps) => (
          <React.Fragment>
            <View style={styles.formHeader}>
              <Text style={styles.title}>Add Product</Text>
              <Picker
                // testID="tax-picker"
                style={{
                  backgroundColor: String(appleBlue),
                  width: 250,
                  fontSize: 15,
                  fontWeight: "bold",
                  color: String(appleBlue),
                  margin: 0,
                  padding: 0,
                }}
                selectedValue={tax}
                onValueChange={(itemValue) => {
                  setTax(itemValue);
                }}>
                {/* hard coded taxes */}
                <Picker.Item label='State Selected: AK - 5.0%' value='0.05' />
                <Picker.Item label='State Selected: AL - 6.0%' value='0.06' />
                <Picker.Item label='State Selected: AR - 3.0%' value='0.03' />
                <Picker.Item label='State Selected: AS - 5.7%' value='0.057' />
                <Picker.Item label='State Selected: AZ - 10.0%' value='0.1' />
              </Picker>
            </View>
            <TextInput
              style={styles.TextInput}
              placeholder='Enter Product Label'
              value={formikProps.values.productLabel}
              onChangeText={formikProps.handleChange("productLabel")}
            />
            <Text style={styles.errorMsg}>
              {formikProps.touched.productLabel &&
                formikProps.errors.productLabel}
            </Text>
            <View
              style={[styles.container, { justifyContent: "space-between" }]}>
              <View
                style={[styles.container, { justifyContent: "flex-start" }]}>
                <View style={[styles.formHeader, { padding: 0 }]}>
                  <TextInput
                    keyboardType='numeric'
                    style={[
                      styles.TextInput,
                      styles.smallTextInput,
                      ,
                      { width: 130 },
                    ]}
                    placeholder='Enter Price'
                    value={formikProps.values.productPrice.toString()}
                    onChangeText={formikProps.handleChange("productPrice")}
                  />
                  <TextInput
                    keyboardType='numeric'
                    style={[
                      styles.TextInput,
                      styles.smallTextInput,
                      { width: 140 },
                    ]}
                    placeholder='Enter Quantity'
                    value={formikProps.values.quantity.toString()}
                    onChangeText={formikProps.handleChange("quantity")}
                  />
                </View>
              </View>
              <View
                style={{
                  display: "flex",
                  flexDirection: "column",
                  justifyContent: "center",
                }}>
                {formikProps.isSubmitting ? (
                  <ActivityIndicator color={appleBlue} />
                ) : (
                  <>
                    <Text
                      onPress={() => {
                        formikProps.handleSubmit();
                      }}
                      style={styles.clickableText}>
                      Add to list
                    </Text>
                  </>
                )}
              </View>
            </View>
            <View
              style={
                !!!formikProps.errors
                  ? { display: "none" }
                  : { display: "flex" }
              }>
              <Text style={styles.errorMsg}>
                {formikProps.touched.productPrice &&
                  formikProps.errors.productPrice}
              </Text>
              <Text style={styles.errorMsg}>
                {formikProps.touched.quantity && formikProps.errors.quantity}
              </Text>
            </View>
          </React.Fragment>
        )}
      </Formik>
    </View> 
  );
};

const styles = StyleSheet.create({
  formHeader: {
    padding: 6,
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    alignContent: "center",
  },
  title: {
    fontSize: 18,
    // fontWeight: "bold",
  },
  clickableText: {
    fontSize: 15,
    fontWeight: "bold",
    color: String(appleBlue),
    alignItems: "center",
    marginRight: 10,
  },
  TextInput: {
    margin: 2,
    marginVertical: 8,
    borderWidth: 1,
    borderColor: String(lightGrey),
    borderRadius: 4,
    padding: 8,
    marginBottom: 3,
    fontSize: 16,
  },
  smallTextInput: {
    width: 120,
    marginRight: 10,
  },
  errorMsg: {
    color: "red",
    fontSize: 12,
  },
  container: {
    display: "flex",
    flexDirection: "row",
  },
  Screenontainer: {
    backgroundColor: "white",
    padding: 10,
    paddingBottom: 0,
  },
});

export default ItemForm;
