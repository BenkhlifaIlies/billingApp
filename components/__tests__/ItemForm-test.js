import React from "react";
import renderer from "react-test-renderer";
import { configure, shallow } from "enzyme";

import ItemForm from "../ItemForm";

import { BillProvider } from "../../providers/BillContext";
import { ProductProvider } from "../../providers/ProductContext";

import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

test("renders correctly", () => {
  const tree = renderer
    .create(
      <ProductProvider>
        <BillProvider>
          <ItemForm />
        </BillProvider>
      </ProductProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
