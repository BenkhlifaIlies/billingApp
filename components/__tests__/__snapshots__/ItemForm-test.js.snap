// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "padding": 10,
      "paddingBottom": 0,
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 6,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      Add Product
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#7ABCDD",
          "color": "#7ABCDD",
          "fontSize": 15,
          "fontWeight": "bold",
          "margin": 0,
          "padding": 0,
          "width": 250,
        }
      }
    >
      <RNCPicker
        items={
          Array [
            Object {
              "label": "State Selected: AK - 5.0%",
              "testID": undefined,
              "textColor": undefined,
              "value": "0.05",
            },
            Object {
              "label": "State Selected: AL - 6.0%",
              "testID": undefined,
              "textColor": undefined,
              "value": "0.06",
            },
            Object {
              "label": "State Selected: AR - 3.0%",
              "testID": undefined,
              "textColor": undefined,
              "value": "0.03",
            },
            Object {
              "label": "State Selected: AS - 5.7%",
              "testID": undefined,
              "textColor": undefined,
              "value": "0.057",
            },
            Object {
              "label": "State Selected: AZ - 10.0%",
              "testID": undefined,
              "textColor": undefined,
              "value": "0.1",
            },
          ]
        }
        onChange={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Enter Product Label"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "#E7E7E7",
        "borderRadius": 4,
        "borderWidth": 1,
        "fontSize": 16,
        "margin": 2,
        "marginBottom": 3,
        "marginVertical": 8,
        "padding": 8,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      Object {
        "color": "red",
        "fontSize": 12,
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
        },
        Object {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flexDirection": "row",
          },
          Object {
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignContent": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 6,
            },
            Object {
              "padding": 0,
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholder="Enter Price"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderColor": "#E7E7E7",
                "borderRadius": 4,
                "borderWidth": 1,
                "fontSize": 16,
                "margin": 2,
                "marginBottom": 3,
                "marginVertical": 8,
                "padding": 8,
              },
              Object {
                "marginRight": 10,
                "width": 120,
              },
              undefined,
              Object {
                "width": 130,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholder="Enter Quantity"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderColor": "#E7E7E7",
                "borderRadius": 4,
                "borderWidth": 1,
                "fontSize": 16,
                "margin": 2,
                "marginBottom": 3,
                "marginVertical": 8,
                "padding": 8,
              },
              Object {
                "marginRight": 10,
                "width": 120,
              },
              Object {
                "width": 140,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "color": "#7ABCDD",
            "fontSize": 15,
            "fontWeight": "bold",
            "marginRight": 10,
          }
        }
      >
        Add to list
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "red",
          "fontSize": 12,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "red",
          "fontSize": 12,
        }
      }
    />
  </View>
</View>
`;
